File: WINGS_AI_Bootcamp/Python/Check_Anagram.py
#   ANAGRAM  - rearranging the letters of another word or phrase, using all the original letters exactly once.
#               Ex: "listen" and "silent"

import re
from collections import Counter

def clean(s: str) -> str:
    return re.sub(r'[^a-z0-9]', '', s.lower())

def is_anagram_clean(s1: str, s2: str) -> bool:
    return Counter(clean(s1)) == Counter(clean(s2))


# OR

def is_anagram(s1: str, s2: str) -> bool:
    return sorted(s1) == sorted(s2)

print(is_anagram("silet", "listen"))
print(is_anagram_clean("tan", "ant"))

File: WINGS_AI_Bootcamp/Haskell/Check_Anagram.hs
import Data.Char (toLower, isAlphaNum)
import Data.List (sort)

-- Clean a string: keep only alphanumeric, convert to lowercase
clean :: String -> String
clean = map toLower . filter isAlphaNum

-- Check if two strings are anagrams after cleaning
isAnagramClean :: String -> String -> Bool
isAnagramClean s1 s2 = sort (clean s1) == sort (clean s2)

-- Simple anagram check (no cleaning)
isAnagram :: String -> String -> Bool
isAnagram s1 s2 = sort s1 == sort s2

main :: IO ()
main = do
    print $ isAnagram "silet" "listen"       -- False, because "silet" vs "listen" differ
    print $ isAnagramClean "tan" "ant"       -- True
    print $ isAnagramClean "Listen!" "Silent" -- True
